src/Vault.sol

```diff
@@ -0,0 +1,261 @@
+    }
+
+    function totalAssets() public view returns (uint256) {
+        return Storage.getTotalAssets();
+    }
+
+    function paused() public view returns (bool) {
+        return Storage.getPaused();
+    }
+
+    function rateProvider() public view returns (address) {
+        return Storage.getRateProvider();
+    }
+
+    function maxDeposit(address) public pure returns (uint256) {
+        return Storage.getMaxDeposit();
+    }
+
+    function maxMint(address) public pure returns (uint256) {
+        return Storage.getMaxMint();
+    }
+
+    // QUESTION: How to handle this in v1 with async withdraws.
+    function maxWithdraw(address owner) public view returns (uint256) {
+        return Storage.convertToAssets(asset(), balanceOf(owner), Math.Rounding.Floor);
+    }
+
+    // QUESTION: How to handle this in v1 with async withdraws.
+    function maxRedeem(address owner) public view returns (uint256) {
+        return balanceOf(owner);
+    }
+
+    function previewDeposit(uint256 assets_) public view returns (uint256) {
+        return Storage.convertToShares(asset(), assets_, Math.Rounding.Floor);
+    }
+
+    function previewDepositAsset(address asset_, uint256 assets_) public view returns (uint256) {
+        return Storage.convertToShares(asset_, assets_, Math.Rounding.Floor);
+    }
+
+    function previewMint(uint256 shares_) public view returns (uint256) {
+        return Storage.convertToAssets(asset(), shares_, Math.Rounding.Ceil);
+    }
+
+    function previewMintAsset(address asset_, uint256 shares_) public view returns (uint256) {
+        return Storage.convertToAssets(asset_, shares_, Math.Rounding.Ceil);
+    }
+
+    function convertToShares(uint256 assets_) public view returns (uint256) {
+        return Storage.convertToShares(asset(), assets_, Math.Rounding.Floor);
+    }
+
+    function convertToAssets(uint256 shares) public view returns (uint256) {
+        return Storage.convertToAssets(asset(), shares, Math.Rounding.Floor);
+    }
+
+    // QUESTION: How to handle this? Start disabled, come back later
+    // This would have to be it's own Liquidity and Risk Module
+    // that calculates the asset ratios and figure out the debt ratio
+    function previewWithdraw(uint256 assets_) public view returns (uint256) {
+        return Storage.convertToShares(asset(), assets_, Math.Rounding.Ceil);
+    }
+
+    // QUESTION: How do we handle this?
+    function previewRedeem(uint256 shares_) public view returns (uint256) {
+        return Storage.convertToAssets(asset(), shares_, Math.Rounding.Floor);
+    }
+
+    function deposit(uint256 assets_, address receiver) public returns (uint256) {
+        if (paused()) revert Paused();
+
+        uint256 shares = previewDeposit(assets_);
+        _deposit(asset(), _msgSender(), receiver, assets_, shares);
+
+        return shares;
+    }
+
+    function depositAsset(address asset_, uint256 assets_, address receiver) public returns (uint256) {
```

Comment:
we may consider overloading here with 
deposit(address asset_, uint256 assets_, address receiver)